### PR TITLE
8261235: C1 compilation fails with assert(res->vreg_number() == index) failed: conversion check

### PR DESCRIPTION
**Pull request by zangcq**
Hi all,

This is backport [JDK-8261235](https://bugs.openjdk.org/browse/JDK-8261235) from jdk17 to jdk8u-dev, to fix C1 compilation failed in HugeMethods.

File 'hotspot/test/compiler/c1/TestTooManyVirtualRegistersMain.java'  remove  annotation '  @requires vm.compiler1.enabled
'
other file mainly in copyrighht year different.  

it's necessary for jdk8u-dev.

Change has been verified locally, we do test jdk-tier1 tier2 and hotspot-tier1 passed, no risk.

### Issue
 [JDK-8261235](https://bugs.openjdk.org/browse/JDK-8261235) : C1 compilation fails with assert(res->vreg_number() == index) failed: conversion check

---------
- [x] I confirm that I make this contribution in accordance with the [OpenJDK Interim AI Policy](https://openjdk.org/legal/ai).

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] [JDK-8261235](https://bugs.openjdk.org/browse/JDK-8261235) needs maintainer approval
- [x] Change must be properly reviewed (2 reviews required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer), 1 [Author](https://openjdk.org/bylaws#author))

### Issue
 * [JDK-8261235](https://bugs.openjdk.org/browse/JDK-8261235): C1 compilation fails with assert(res-&gt;vreg_number() == index) failed: conversion check (**Bug** - P3 - Approved)

### Reviewers
 * [Christian Hagedorn](https://openjdk.org/census#chagedorn) (@chhagedorn - no project role) Review applies to [c44095fd](https://git.openjdk.org/jdk8u-dev/pull/622/files/c44095fd9413d37350b2d7fd76383d0c0badd1e9)
 * [Roland Westrelin](https://openjdk.org/census#roland) (@rwestrel - **Reviewer**)
 * [Sergey Bylokhov](https://openjdk.org/census#serb) (@mrserb - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk8u-dev.git pull/622/head:pull/622` \
`$ git checkout pull/622`

Update a local copy of the PR: \
`$ git checkout pull/622` \
`$ git pull https://git.openjdk.org/jdk8u-dev.git pull/622/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 622`

View PR using the GUI difftool: \
`$ git pr show -t 622`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk8u-dev/pull/622.diff">https://git.openjdk.org/jdk8u-dev/pull/622.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk8u-dev/pull/622#issuecomment-2710938422)
</details>

### CODE REVIEW

**Comment by zangcq** [comment]
Hi @chhagedorn , I will backport your patch to jdk8 , please review

File: hotspot/src/share/vm/c1/c1_LinearScan.cpp [hunk 1]
```diff
@@ -3857,12 +3857,27 @@ void MoveResolver::insert_move(LIR_Opr from_opr, Interval* to_interval) {
   assert(_insert_list != NULL && _insert_idx != -1, "must setup insert position first");
   assert(_insertion_buffer.lir_list() == _insert_list, "wrong insertion buffer");
 
-  LIR_Opr to_opr = LIR_OprFact::virtual_register(to_interval->reg_num(), to_interval->type());
+  LIR_Opr to_opr =get_virtual_register(to_interval);
```
Comment (chhagedorn) [inline hotspot/src/share/vm/c1/c1_LinearScan.cpp:3860]:
```suggestion
  LIR_Opr to_opr = get_virtual_register(to_interval);
```

**Review by chhagedorn — APPROVED** [review]
Otherwise, looks good to me!

File: hotspot/src/share/vm/c1/c1_LinearScan.cpp (hunk 1 above)
Comment (zangcq) [reply hotspot/src/share/vm/c1/c1_LinearScan.cpp]: thanks for your suggestion, Fixed :-)

**Review by chhagedorn — APPROVED** [review]

**Comment by zangcq** [comment]
/approval  8261235  request  C1 compilation fails with assert(res->vreg_number() == index) failed: conversion check

**Comment by zangcq** [comment]
Kindly ping @gnu-andrew , could you please review this patch?  hagedorn just have read-only permissions.

**Comment by jerboaa** [comment]
@zangcq Please enable GHA on your fork.

**Comment by zangcq** [comment]
> Why is this an important fix to have in OpenJDK 8u? Why does the risk of adding it at this stage of the lifecycle of 8u not outweigh the benefits?

Thanks for your reply @jerboaa . 

1.  As I mentioned in "C1 compilation failed in HugeMethods", I explained the background of our specific case. In our Flink datastream program, we discovered that numerous hot methods generated by jsoniter or fastjson were being interpreted. The bytecode of these methods was incorrectly generated and exceeded the compilation threshold. When we added the -XX:-DontCompileHugeMethods option to the command line, the JDK 8u consistently crashed. To reproduce the identical stack backtrace, we built the fastdebug version, and with this pull request (pr), the crash disappear. This is why we believe this issue is of significant importance to OpenJDK 8u.​

2.  this pr only to check  vreg_number in linear scan stage to bailout the compile task.  and there is a follow fix  https://bugs.openjdk.org/browse/JDK-8287396 . because 8u and 17u with different c++ standards, the follow pr build failed in OpenJDK 8u.

**Comment by zangcq** [comment]
> @zangcq Please enable GHA on your fork.

done

**Comment by zangcq** [comment]
> @zangcq Please enable GHA on your fork.

@jerboaa

**Comment by zangcq** [comment]
> @zangcq Please enable GHA on your fork.

@jerboaa FYI：GHA Result
- with this pr： https://github.com/zangcq/jdk8u-dev/actions/runs/15105864713  
 Only security releated case failed，`TestTooManyVirtualRegistersMain`  in hotspot/tier1 passed

- without this pr ：https://github.com/zangcq/jdk8u-dev/actions/runs/15105793305
 Beside security releated case failed，`TestTooManyVirtualRegistersMain`  failed.

```
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  SIGSEGV (0xb) at pc=0x00007fda89da9f99, pid=43889, tid=0x00007fda62afd640
#
# JRE version: OpenJDK Runtime Environment (8.0_462) (build 1.8.0_462-internal-zangcq-64a3d592291d15076edac82a975e797f4c327341-b00)
# Java VM: OpenJDK 64-Bit Server VM (25.462-b00 mixed mode linux-amd64 compressed oops)
# Problematic frame:
# V  [libjvm.so+0x3a9f99]  LinearScan::eliminate_spill_moves()+0x1b9
#
# Failed to write core dump. Core dumps have been disabled. To enable core dumping, try "ulimit -c unlimited" before starting Java again
#
# An error report file with more information is saved as:
# /home/runner/work/jdk8u-dev/jdk8u-dev/test-results/testoutput/hotspot_tier1/JTwork/scratch/hs_err_pid43889.log
#
# Compiler replay data is saved as:
# /home/runner/work/jdk8u-dev/jdk8u-dev/test-results/testoutput/hotspot_tier1/JTwork/scratch/replay_pid43889.log
#
# If you would like to submit a bug report, please visit:
#   http://bugreport.java.com/bugreport/crash.jsp
#
```
my inside log build with fastdebug mode 

```
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  Internal Error (jdk8u-dev/hotspot/src/share/vm/c1/c1_LIR.hpp:744), pid=4163761, tid=0x00007ffb160a2700
#  assert(res->vreg_number() == index) failed: conversion check
#
# JRE version: OpenJDK Runtime Environment (8.0_462) (build 1.8.0_462-fastdebug-zangchuanqi_2025_05_19_16_16-b00)
# Java VM: OpenJDK 64-Bit Server VM (25.462-b00-fastdebug mixed mode linux-amd64 compressed oops)
# Failed to write core dump. Core dumps have been disabled. To enable core dumping, try "ulimit -c unlimited" before starting Java again
#
# If you would like to submit a bug report, please visit:
#   http://bugreport.java.com/bugreport/crash.jsp
#

---------------  T H R E A D  ---------------

Current thread (0x00007ffb2c183000):  JavaThread "C1 CompilerThread3" daemon [_thread_in_native, id=4163781, stack(0x00007ffb15fa3000,0x00007ffb160a3000)]

Stack: [0x00007ffb15fa3000,0x00007ffb160a3000],  sp=0x00007ffb160a0160,  free space=1012k
Native frames: (J=compiled Java code, j=interpreted, Vv=VM code, C=native code)
V  [libjvm.so+0x13457db]  VMError::report_and_die()+0x21b
V  [libjvm.so+0x82bbe1]  report_vm_error(char const*, int, char const*, char const*)+0x61
V  [libjvm.so+0x570459]  LIR_OprFact::virtual_register(int, BasicType)+0x89
V  [libjvm.so+0x58cedd]  MoveResolver::insert_move(Interval*, Interval*)+0xcd
V  [libjvm.so+0x5a5c4b]  MoveResolver::resolve_mappings()+0x11b
V  [libjvm.so+0x5a6279]  MoveResolver::move_insert_position(LIR_List*, int)+0x69
V  [libjvm.so+0x5a6583]  LinearScanWalker::insert_move(int, Interval*, Interval*)+0x283
V  [libjvm.so+0x5a6bbe]  LinearScanWalker::split_for_spilling(Interval*)+0x58e
V  [libjvm.so+0x5a6f22]  LinearScanWalker::split_and_spill_intersecting_intervals(int, int)+0x62
V  [libjvm.so+0x5a759a]  LinearScanWalker::alloc_locked_reg(Interval*)+0x58a
V  [libjvm.so+0x5a797d]  LinearScanWalker::activate_current()+0x24d
V  [libjvm.so+0x5965c1]  IntervalWalker::walk_to(int)+0x611
V  [libjvm.so+0x5a8f4c]  LinearScan::allocate_registers()+0x25c
V  [libjvm.so+0x5ad37a]  LinearScan::do_linear_scan()+0x14a
V  [libjvm.so+0x4d9e6f]  Compilation::emit_lir()+0x99f
V  [libjvm.so+0x4dc57e]  Compilation::compile_java_method()+0x47e
V  [libjvm.so+0x4dce94]  Compilation::compile_method()+0x64
V  [libjvm.so+0x4dd664]  Compilation::Compilation(AbstractCompiler*, ciEnv*, ciMethod*, int, BufferBlob*)+0x314
V  [libjvm.so+0x4de154]  Compiler::compile_method(ciEnv*, ciMethod*, int)+0x144
V  [libjvm.so+0x78f015]  CompileBroker::invoke_compiler_on_method(CompileTask*)+0x5b5
V  [libjvm.so+0x7911e8]  CompileBroker::compiler_thread_loop()+0x928
V  [libjvm.so+0x12b090e]  JavaThread::thread_main_inner()+0x25e
V  [libjvm.so+0x1037430]  java_start(Thread*)+0x140
C  [libpthread.so.0+0x74a4]  start_thread+0xc4


Current CompileTask:
C1:   4317   34    b  3       compiler.c1.TestTooManyVirtualRegisters::test (7326 bytes)

```

**Comment by jerboaa** [comment]
@rwestrel Could you please help get a second review of this? Thank you.

**Review by rwestrel — APPROVED** [review]
Looks good to me.

**Comment by zangcq** [comment]
/approval request Fix When jdk8u added the -XX:-DontCompileHugeMethods to compile a hugemethod，it is a corner case in C1, Patch applies cleanly to 8u and add a testcase. New regression test only fails jdk/security_infra. Patched JDK passes tier1.

**Comment by zangcq** [comment]
/approval request Fix  crash when jdk8u added the -XX:-DontCompileHugeMethods to compile a hugemethod, it is a corner case in C1. Patch applies cleanly to 8u and add a testcase. New regression test only fails in jdk/security_infra. Patched JDK passes tier1

**Comment by jerboaa** [comment]
See the progress checklist. You need another review from at least an author in jdk8u.

**Comment by zangcq** [comment]
Hi @mrserb,  cloud you please review this patch?

**Review by mrserb — APPROVED** [review]

File: hotspot/test/compiler/c1/TestTooManyVirtualRegistersMain.java [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+/*
+ * @test
+ * @bug 8261235
+ * @summary Tests custom bytecode which requires too many virtual registers in the linear scan of C1.
+ *          The test should bail out in C1.
+ * @compile TestTooManyVirtualRegisters.jasm
+ * @run main/othervm -Xbatch -XX:CompileCommand=dontinline,compiler.c1.TestExceptionBlockWithPredecessors::*
+ *                   compiler.c1.TestTooManyVirtualRegistersMain
+ */
```
Comment (mrserb) [inline hotspot/test/compiler/c1/TestTooManyVirtualRegistersMain.java:32]: Note that there is a typo here, the wrong name is used for CompileCommand.
I have filed: https://bugs.openjdk.org/browse/JDK-8381880

Comment (zangcq) [reply hotspot/test/compiler/c1/TestTooManyVirtualRegistersMain.java:32]: Appreciate it for your comments，should we fix this typo in this pr？

Comment (mrserb) [reply hotspot/test/compiler/c1/TestTooManyVirtualRegistersMain.java:32]: no that is a separate issue, and it does not affect the ability to reproduce the bug.

**Comment by jerboaa** [comment]
@zangcq You need to issue `/template append` and confirm the AI policy before this can be integrated.

**Comment by zangcq** [comment]
/template append  - [x] I confirm that I make this contribution in accordance with the [OpenJDK Interim AI Policy](https://openjdk.org/legal/ai)

**Comment by zangcq** [comment]
> @zangcq You need to issue `/template append` and confirm the AI policy before this can be integrated.

Thanks for your advice，done
